### ISSUE TITLE
GVim crash on call to "setreg"

### ISSUE THREAD

**Issue by sosmo** [vim/vim#5182]
**Describe the bug**
When calling setreg under specific conditions, gVim crashes with the following output to stderr:

```
BadWindow (invalid Window parameter)
Vim: Got X error
Vim: Finished.
```

**To Reproduce**
1. Run `gvim --u NONE`
2. Copy the text in the paste to the system clipboard https://pastebin.com/0aUzkDYX
3. Run the following ex commands:
```
let content = getreg('+')
let content = substitute(content, '\n$', '', '')
call setreg('+', content, 'c')
```

The last ex command will cause a crash

**Environment**
 - Vim version: 8.1, patches 1-320
 - OS: Ubuntu 19.04, KDE Plasma 5.15 (X version)
 - Terminal: GUI (vim-gtk)

**Comment by brammool** [comment]
I cannot reproduce.  Can you try a newer version of Vim?
If you use some kind of cllipboard manager, try disabling it.

**Comment by chrisbra** [comment]
cannot reproduce with 8.1.2225, gtk2 and xfce. please try to reproduce with a newer vim.

**Comment by sosmo** [comment]
Thanks for the answers. I'll test on a newer build soon, Ubuntu's Vim seems to be woefully outdated. If it doesn't get resolved I'll open another issue.

**Comment by jlittlenz** [comment]
This has been reported before. It is issue #1023, from 2016, but I commented then that I'd struck it years before, with vim 7.  I can reproduce this every time.  I've no idea why it only affects some users.
The X error occurs when some other programme, typically a clipboard manager, requests the selection text from vim, and the text is large enough.  This makes it hard to debug, as the request arrives asynchronously.
I tried hard to track this down in a debugger, but the X error originates within GTK code somewhere, and vim hasn't done anything different than when the text is smaller and the error doesn't occur.  I concluded that it's an old GTK bug and gave up.
The workaround I have used for many years is to tell vim to ignore the error.  In the function x_error_handler in os_unix.c I comment out the preserve_exit() call.  A comment there suggested this: "Silently continuing might be an alternative...".  AFAICT I've had no negative consequences from ignoring these X errors.  I still see the error reported if I start gvim from a terminal, and they don't occur often, except if I select a lot of text.
Configuring vim to not always acquire the X selection avoids the problem, but I found that a nuisance.
